subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1065 @@
+
+	struct audio_module_message *msg_rx;
+	size_t msg_rx_size;
+
+	if (handle == NULL) {
+		LOG_DBG("Module handle is NULL");
+		return -EINVAL;
+	}
+
+	if (handle->state != AMOD_STATE_RUNNING ||
+	    handle->description->type == AMOD_TYPE_UNDEFINED ||
+	    handle->description->type == AMOD_TYPE_OUTPUT) {
+		LOG_DBG("Module %s in an invalid state (%d) or type (%d) to receive data",
+			handle->name, handle->state, handle->description->type);
+		return -ENOTSUP;
+	}
+
+	if (block == NULL || block->data == NULL || block->data_size == 0) {
+		LOG_DBG("Error in audio block for module %s", handle->name);
+		return -ECONNREFUSED;
+	}
+
+	ret = data_fifo_pointer_last_filled_get(handle->thread.msg_rx, (void **)&msg_rx,
+						&msg_rx_size, timeout);
+	if (ret) {
+		LOG_DBG("Failed to retrieve data from module %s, ret %d", handle->name, ret);
+		return ret;
+	}
+
+	if (msg_rx->block.data == NULL || msg_rx->block.data_size > block->data_size) {
+		LOG_DBG("Data output buffer NULL or too small for received buffer from module %s",
+			handle->name);
+		ret = -EINVAL;
+	} else {
+		memcpy(&block->meta, &msg_rx->block.meta, sizeof(struct audio_metadata));
+		memcpy((uint8_t *)block->data, (uint8_t *)msg_rx->block.data,
+		       msg_rx->block.data_size);
+	}
+
+	block_release_cb((struct audio_module_handle_private *)handle, &msg_rx->block);
+
+	data_fifo_block_free(handle->thread.msg_rx, (void **)&msg_rx);
+
+	return ret;
+}
+
+/**
+ * @brief Send an audio data block to a module and retrieve an audio data block from a module.
+ *
+ * @note The block is processed within the module or sequence of modules. The result is returned
+ *       via the module or final module's output FIFO.
+ *       All the input data is consumed within the call and thus the input data buffer
+ *       maybe released once the function returns.
+ *
+ */
+int audio_module_data_tx_rx(struct audio_module_handle *handle_tx,
+			    struct audio_module_handle *handle_rx, struct audio_data *block_tx,
+			    struct audio_data *block_rx, k_timeout_t timeout)
+{
+	int ret;
+	struct audio_module_message *msg_rx;
+	size_t msg_rx_size;
+
+	if (handle_tx == NULL || handle_rx == NULL) {
+		LOG_DBG("Module handle is NULL");
+		return -EINVAL;
+	}
+
+	if (handle_tx->state != AMOD_STATE_RUNNING || handle_rx->state != AMOD_STATE_RUNNING) {
+		LOG_DBG("Module is in an invalid state or type to send-receive data");
+		return -ENOTSUP;
+	}
+
+	if (handle_tx->description->type == AMOD_TYPE_UNDEFINED ||
+	    handle_tx->description->type == AMOD_TYPE_INPUT ||
+	    handle_rx->description->type == AMOD_TYPE_UNDEFINED ||
+	    handle_rx->description->type == AMOD_TYPE_OUTPUT) {
+		LOG_DBG("Module not of the right type for operation");
+		return -ENOTSUP;
+	}
+
+	if (block_tx == NULL || block_rx == NULL) {
+		LOG_DBG("Block pointer for module is NULL");
+		return -ECONNREFUSED;
+	}
+
+	if (block_tx->data == NULL || block_tx->data_size == 0 || block_rx->data == NULL ||
+	    block_rx->data_size == 0) {
```

Comment:
I thought the rx would be 0 and then populated by this function? Ref. our discussuion.

Reply:
Can we discuss again?
As discussed the size is the number of bytes within the buffer and this test is to ensure we have memory to write to.

Reply:
Not important for now.